src/lib/common/sol-platform-impl-linux-micro.c

```diff
@@ -514,16 +521,108 @@ teardown_pid1(void)
 
+static bool
+sol_platform_linux_micro_should_debug(struct gdbserver *gdb)
+{
+    int idx, argc = sol_argc();
+    char **argv = sol_argv();
+    bool res = false;
+
+    for (idx = 1; idx < argc; idx++) {
+        if (streq(argv[idx], SOL_DEBUG_ARG))
+            res = true;
+        else if (strstartswith(argv[idx], SOL_DEBUG_COMM_ARG)) {
+            gdb->comm = argv[idx] + strlen(SOL_DEBUG_COMM_ARG);
+            if (res) {
+                SOL_DBG("Got "SOL_DEBUG_ARG" flag and comm");
+                return res;
+            }
+        }
+    }
+
+    return res;
+}
+
+static void
+on_fork(void *data)
+{
+    int err;
+    unsigned int i;
+    char pids[PATH_MAX];
+    const char *gdbserver = NULL;
+    struct gdbserver *gdb = data;
+
+    const char *paths[] = {
+        "/usr/bin/gdbserver",
+        "/bin/gdbserver",
+    };
+
+    for (i = 0; i < ARRAY_SIZE(paths); i++) {
+        struct stat st;
+        err = stat(paths[i], &st);
+        if (err < 0) {
+            if (errno == ENOENT) {
+                SOL_DBG("Could not find gdbserver executable at: %s", paths[i]);
+                continue;
+            } else {
+                SOL_DBG("Could not stat gdbserver executable at: %s - %s",
+                        paths[i], sol_util_strerrora(errno));
+                continue;
+            }
+        } else if (!(st.st_mode & S_IXUSR)) {
+            SOL_DBG("gdbserver executable at: %s - is not executable", paths[i]);
+            sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+            continue;
+        }
+
+        gdbserver = paths[i];
+        break;
+    }
+
+    if (!gdbserver) {
+        SOL_DBG("Could not find any gdbserver executable installed");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    err = snprintf(pids, sizeof(pids), "%d", gdb->pid);
+    if (err < 0 || err >= (int)sizeof(pids)) {
+        SOL_ERR("Could not format pid string to attach gdbserver");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    if (execl(gdbserver, gdbserver, "--attach", gdb->comm, pids, (char *)0) == -1) {
+        SOL_ERR("failed to exec gdbserver - %s", sol_util_strerrora(errno));
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    SOL_ERR("Failed to exec() gdbserver");
+    sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+}
+
+static void
+on_fork_exit(void *data, uint64_t pid, int status)
+{
+    SOL_ERR("gdbserver exited... %"PRIu64, pid);
+}
+
 int
 sol_platform_impl_init(void)
 {
     bool want_load_initial_services = false;
+    struct gdbserver gdb;
+    pid_t pid = getpid();
 
-    if (getpid() == 1 && getppid() == 0) {
+    if (pid == 1 && getppid() == 0) {
         int err = setup_pid1();
         SOL_INT_CHECK(err, < 0, err);
 
         want_load_initial_services = true;
+
+        gdb.pid = pid;
+        gdb.comm = "/dev/tty0";
+
+        if (sol_platform_linux_micro_should_debug(&gdb))
+            sol_platform_linux_fork_run(on_fork, on_fork_exit, &gdb);
```

Comment:
hum... I didn't test, but I did think it would not work. Did you confirm it's working?

Usually if you're PID1 you have some restrictions and I thought PTRACE would be forbidden, also not sure what happens if we crash it somehow.

What I initially envisioned was to keep PID1 as a very slim process that spawned the gdbserver and soletta application itself. It would `setenv("SOL_LOAD_INITIAL_SERVICES=1")` to force the application to do its stuff.

We need to be careful to have this slim process to setup the system, such as mounting `/proc`, `/dev`... (that's what `setup_pid1()` does) but not more than that, otherwise we may lose important code paths (ie: one may want  to debug a linux-micro initial service).

The other bit to do in the slim process is to handle SIGCHLD to `waitpid()`... as before, we should reduce code paths, thus no use of main loop stuff, just do a loop with `waitpid()` and it should be good.

summary:

```c
if (pid == 1 && getppid() == 0) {
   int err = setup_pid1();
   SOL_INT_CHECK(err, < 0, err);

   want_load_initial_services = true;
   gdb.comm =  "/dev/tty0";
   if (should_debug(&gdb)) {
      pid_t gdbserver_pid = fork();
      if (gdbserver_pid == 0) {
          args = alloca((_argc + 2) * sizeof(char*));
          args[0] = gdbserver_path;
          args[1] = gdb.comm;
          memcpy(args + 2, _argv, _argc * sizeof(char *));
          setenv("SOL_LOAD_INITIAL_SERVICES=1"); // so the re-execution will still run the services
          execv(gdbserver_path, args); // replace this with a loop that tries all paths. call alloca() BEFORE the loop!
          // does not return. should exit(EXIT_FAILURE)
      } else if (gdbserver_pid > 0) {
          // we enter a minimal system, do not return, just keep collecting zombie processes that goes to pid1
          while (true) {
             int status;
             pid_t child = wait(&status);
             // error handling
            SOL_DBG("child pid=%"PRIu64" exit=%d", (uint64_t)child, status);
          }
          // does not return
      } else { // error handling. I'd just SOL_WRN() and keep going without gdbserver }
   }
}
```

Another reason to do so is to avoid race conditions. If you call `gdbserver --attach PID`, then you may lose important debug paths while gdbserver starts. The only way to avoid them is to have gdbserver to start the process. Then what we do is to re-exec by taking our argv.

The trick is to set `SOL_LOAD_INITIAL_SERVICES=1` so the application will still execute linux micro services even not being pid1.